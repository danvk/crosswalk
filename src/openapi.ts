import * as pathToRegexp from 'path-to-regexp';

type Schema = {$ref: string};

interface BodyParam {
  in: 'body';
  name: string;
  schema: Schema;
}

interface PathParam {
  in: 'path';
  name: string;
  type: 'string';
  description?: string;
}

interface QueryParam {
  in: 'query';
  name: string;
  schema: Schema;
  description?: string;
}

type Param = PathParam | QueryParam | BodyParam;

interface SwaggerEndpoint {
  summary?: string;
  description?: string;
  parameters?: Param[];
  responses: {
    [statusCode: string]: {
      description?: string;
      schema: Schema;
    };
  };
}

export interface Options {
  info?: {
    title?: string;
    description?: string;
    version?: string;
  };
  host?: string;
  basePath?: string;
  schemes?: ('http' | 'https')[];
}

const isObject = <T>(x: T): x is object & T => !!x && typeof x === 'object';

function extractPathParams(path: string): PathParam[] {
  const tokens = pathToRegexp.parse(path);
  return tokens.filter(isObject).map(tok => ({
    name: '' + tok.name,
    in: 'path',
    type: 'string',
  }));
}

/** Convert /foo/:bar/:baz --> /foo/{bar}/{baz} */
function expressPathToOpenApiPath(path: string): string {
  return path.replace(/:([^/]+)/g, '{$1}');
}

const DEFINITION = '#/definitions/';
export function followApiRef(spec: any, schema: Schema): [string, unknown] {
  const ref = schema.$ref;
  if (!ref.startsWith(DEFINITION)) {
    throw new Error(`Confused by ${schema} / ${ref}`);
  }

  const name = ref.slice(DEFINITION.length);
  const def = spec.definitions[name];
  if (!def) {
    throw new Error(`Unable to find definition for ${name}`);
  }
  return [name, def];
}

/** Convert an API spec (generated via typescript-json-schema) to OpenAPI. */
export function apiSpecToOpenApi(apiSpec: any, options?: Options): any {
  apiSpec = JSON.parse(JSON.stringify(apiSpec)); // defensive copy
  const {required: endpoints, properties: endpointSpecs, definitions} = apiSpec;

  // Remove endpoints, helpers
  const paths: Record<string, any> = {};

  const toDelete = new Set<string>();

  for (const endpoint of endpoints) {
    const openApiPath = expressPathToOpenApiPath(endpoint);
    paths[openApiPath] = {};
    const byVerb = endpointSpecs[endpoint].properties;
    for (const [verb, ref] of Object.entries(byVerb)) {
      const [name, schema] = followApiRef(apiSpec, ref as Schema);
      const {request, response, query} = (schema as any).properties;

      const parameters: Param[] = extractPathParams(endpoint);
      if (request?.type !== 'null') {
        parameters.push({
          name: 'body',
          in: 'body',
          schema: request,
        });
      }

      if (query.properties) {
        for (const [key, value] of Object.entries<Schema>(query.properties)) {
          parameters.push({
            name: key,
            in: 'query',
            schema: value,
          });
        }
      }

      const swagger: SwaggerEndpoint = {
        summary: (ref as any).description,
        ...(parameters.length && {parameters}),
        responses: {
          // TODO: do I need to break this down by status?
          200: {
            schema: response,
          },
        },
      };
      paths[openApiPath][verb] = swagger;
      toDelete.add(name);
    }
  }

  toDelete.forEach(name => {
    delete definitions[name];
  });

  return {
    swagger: '2.0',
    info: {
      title: 'Generated API',
      description: 'testing testing',
      version: '',
    },
    paths,
    definitions,
    ...options,
  };
}
