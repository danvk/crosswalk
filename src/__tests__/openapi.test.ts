import {apiSpecToOpenApi} from '../openapi';
import apiSchemaJson from './api.schema.json';

describe('Open API integration', () => {
  const openApi = apiSpecToOpenApi(apiSchemaJson);

  it('should have the expected endpoints', () => {
    expect(openApi).toMatchSnapshot('open-api');
  });

  // TODO:
  // - test that descriptions get set
});
