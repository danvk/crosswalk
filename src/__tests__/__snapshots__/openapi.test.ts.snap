// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Open API integration should have the expected endpoints: open-api 1`] = `
Object {
  "definitions": Object {
    "Partial<Pick<User,\\"name\\"|\\"age\\">>": Object {
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "name": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "Pick<User,\\"name\\"|\\"age\\">": Object {
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "name": Object {
          "type": "string",
        },
      },
      "required": Array [
        "age",
        "name",
      ],
      "type": "object",
    },
    "User": Object {
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "id": Object {
          "type": "string",
        },
        "name": Object {
          "type": "string",
        },
      },
      "required": Array [
        "age",
        "id",
        "name",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "description": "testing testing",
    "title": "Generated API",
    "version": "",
  },
  "paths": Object {
    "/complex": Object {
      "post": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "schema": Object {
              "properties": Object {
                "user": Object {
                  "anyOf": Array [
                    Object {
                      "$ref": "#/definitions/User",
                    },
                    Object {
                      "type": "null",
                    },
                  ],
                },
              },
              "required": Array [
                "user",
              ],
              "type": "object",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": undefined,
      },
    },
    "/users": Object {
      "get": Object {
        "responses": Object {
          "200": Object {
            "schema": Object {
              "properties": Object {
                "users": Object {
                  "items": Object {
                    "$ref": "#/definitions/User",
                  },
                  "type": "array",
                },
              },
              "required": Array [
                "users",
              ],
              "type": "object",
            },
          },
        },
        "summary": "Get the full list of users",
      },
      "post": Object {
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "schema": Object {
              "$ref": "#/definitions/Pick<User,\\"name\\"|\\"age\\">",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": "Create a new user",
      },
    },
    "/users/{userId}": Object {
      "delete": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "type": "string",
          },
          Object {
            "in": "body",
            "name": "body",
            "schema": Object {
              "properties": Object {},
              "type": "object",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": undefined,
      },
      "get": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "type": "string",
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": undefined,
      },
      "patch": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "type": "string",
          },
          Object {
            "in": "body",
            "name": "body",
            "schema": Object {
              "$ref": "#/definitions/Partial<Pick<User,\\"name\\"|\\"age\\">>",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": "Edit an existing user",
      },
      "put": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "type": "string",
          },
          Object {
            "in": "body",
            "name": "body",
            "schema": Object {
              "properties": Object {
                "age": Object {
                  "type": "number",
                },
                "name": Object {
                  "type": "string",
                },
              },
              "type": "object",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/User",
            },
          },
        },
        "summary": undefined,
      },
    },
  },
  "swagger": "2.0",
}
`;
